contracts/SolverTrampoline.sol

```diff
@@ -17,17 +23,41 @@ contract SolverTrampoline {
     /// @dev The domain separator for signing EIP-712 settlements.
     bytes32 public immutable domainSeparator;
 
+    /// @dev Nonce by solver address.
+    mapping(address => uint256) public nonces;
+
     constructor(Settlement settlementContract_) {
         settlementContract = settlementContract_;
         solverAuthenticator = settlementContract_.authenticator();
-        
+
         domainSeparator = keccak256(abi.encode(
             keccak256("EIP712Domain(uint256 chainId,address verifyingContract)"),
             block.chainid,
             address(this)
         ));
     }
 
+    /// @dev Executes a settlement on behalf of a solver.
+    function settle(bytes calldata settlement, uint256 nonce, bytes32 r, bytes32 s, uint8 v) external {
+        bytes32 messageDigest = settlementMessage(settlement, nonce);
+        address solver = ecrecover(messageDigest, v, r, s);
+        if (solver == address(0) || !solverAuthenticator.isSolver(solver)) {
+            revert Unauthorized();
```

Comment:
nit: given ecrecover mismatches are very hard to debug maybe it would be convenient to emit in the error which address was recovered (and is considered un-authorized)

Reply:
This was the motivation for including the `settlementMessage` function, so you could locally do a `ethers.utils.recoverSigner(await solverTrampoline.settlementMessage(...), { r, s, v })`. That being said, I think that putting the address in the error is a 0-cost way to make this more debuggable and a nice suggestion.